internal/pally/registry.go

```diff
@@ -0,0 +1,250 @@
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/model"
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+// A Registry is a collection of metrics, usually scoped to a single package or
+// object. Each Registry is also its own http.Handler, and can serve
+// Prometheus-flavored text and protocol buffer pages for metrics introspection
+// or scraping.
+type Registry struct {
+	http.Handler
+
+	metricsMu   sync.RWMutex
+	metrics     []metric
+	constLabels Labels
+
+	// TODO: To serve our own Prometheus endpoints, we only need to implement
+	// prometheus.Gatherer. The vanilla Prometheus registry is a convenient
+	// implementation, but it enforces metric uniqueness via FNV64a; if we run
+	// into hash collisions in our more-demanding applications, we should use a
+	// simple map[string]struct{} instead.
+	prom      *prometheus.Registry
+	federated []prometheus.Registerer
+
+	pushing atomic.Bool
+	scope   tally.Scope
+	ticker  *time.Ticker
+	stop    chan struct{}
+	stopped chan struct{}
+}
+
+// A RegistryOption configures a Registry.
+type RegistryOption func(*Registry)
+
+// Federated links a pally.Registry with a prometheus.Registerer, so that all
+// metrics created in one also appear in the other.
+func Federated(prom prometheus.Registerer) RegistryOption {
+	return func(r *Registry) {
+		r.federated = append(r.federated, prom)
+	}
+}
+
+// Labeled adds constant labels to a Registry. All metrics created by a
+// Registry inherit its constant labels.
+func Labeled(ls Labels) RegistryOption {
+	return func(r *Registry) {
+		for k, v := range ls {
+			if !model.LabelName(k).IsValid() || !model.LabelValue(v).IsValid() {
+				continue
+			}
+			if !isValidTallyString(k) || !isValidTallyString(v) {
+				continue
+			}
+			r.constLabels[k] = v
+		}
+	}
+}
+
+// NewRegistry constructs a new Registry.
+func NewRegistry(opts ...RegistryOption) *Registry {
+	prom := prometheus.NewRegistry()
+	handler := promhttp.HandlerFor(prom, promhttp.HandlerOpts{
+		ErrorHandling: promhttp.HTTPErrorOnError, // 500 on errors
+	})
+
+	r := &Registry{
+		Handler:     handler,
+		metrics:     make([]metric, 0, _defaultVectorSize),
+		constLabels: make(Labels),
+		prom:        prom,
+		// Assume that we'll be federated with the global prometheus Registry.
+		federated: make([]prometheus.Registerer, 0, 1),
+		stop:      make(chan struct{}),
+		stopped:   make(chan struct{}),
+	}
+	for _, opt := range opts {
+		opt(r)
+	}
+	return r
+}
+
+// Push starts a goroutine that periodically exports all registered metrics to
+// a Tally scope. Each Registry can only push to a single Scope; calling Push a
+// second time returns an error.
+func (r *Registry) Push(scope tally.Scope, tick time.Duration) (context.CancelFunc, error) {
+	if r.pushing.Swap(true) {
+		return nil, errors.New("already pushing to Tally")
+	}
+	r.scope = scope
+	r.ticker = time.NewTicker(tick)
```

Comment:
Optional: scope and ticker are only used from this goroutine, and they're only
set after Push is called the first time. It will be very easy for someone to
accidentally break this by accessing them elsewhere.

Encapsulating the context of the goroutine in a struct that you don't hold
onto would solve this.

```
type pusher struct {
  scope tally.Scope
  ticker *time.Ticker
  // ..
}

func (r *Registry) Push(..) .. {
  if r.pushing.Swap(true) {
    // ..
  }

  pusher := newPusher(..)
  go pusher.Start()
  return pusher.Stop, nil
}
```